Add python_version to template config file

KSerrania opened the pull request:
### What does this PR do?

Adds the `python_version` configuration option to the template configuration file.

### Motivation

Starting with 6.14, users will be able to switch python versions. The configuration file should show this option.

### Additional Notes

Open to any ideas to improve wording and / or placement in the file.
I put it in the "Agent" part of the file, which means it will only appear in the configuration file of the base Agent, not in the `DCA`'s nor `dogstatsd`'s config file.

KSerrania committed:
Add python_version to template config file

olivielpeau reviewed (commented):
Thanks, it was definitely an oversight to not document the option in the config template 👍 

A couple of comments:

* the supported values depend on how the Agent was built. Typically when we start building an Agent with only python 3, the config option will default to 3 (and it probably does not even make sense to document it since that'll be the only allowed value for the option). Let's find a way to handle this here. Also, let's ask the Containers team whether it's an issue to have this option documented in the docker image as well, even though the docker image will only ever include one of python 2 or 3.
* It definitely makes sense to include this in the `Agent` section, since `dogstatsd` and `DCA` don't include python at all. That said, the `python_version` should not be considered and "advanced configuration" option (it's a basic config option), so I'd put it above the beginning of the `Advanced Configuration` section.

KSerrania committed:
Render python_version option only if both python versions are present

If both versions are present, the --with-both-python option is passed
to inv agent.build, which will generate the correct configuration.

The generation is probably suboptimal: we generate datadog.yaml 3 times,
because we first generate every config (so one for py3only, another one
for py2py3), and then we specifically generate the config we were asked
to generate (either one or the other).
This was done because with Go regexes, there's no easy way to exclude
lines that contain a particular word, and I didn't want to list all other
generated config files (that'd mean having to add it here when we add a
new config template).

KSerrania commented on tasks/agent.py:
```diff
@@ -159,9 +159,17 @@ def build(ctx, rebuild=False, race=False, build_include=None, build_exclude=None
+    else:
+        cmd = "go generate -x --run=\"agent-py3\" {}/cmd/agent"
+        ctx.run(cmd.format(REPO_PATH), env=env)
+
```
This is suboptimal, since we generate the `datadog.yaml` file three times:
- twice during the first `go generate`, which runs all `go:generate` directives, including `agent-py3` and `agent-py2py3`,
- plus one when we generate specifically the configuration that was requested.

To prevent this, during the first `go generate` we could exclude the directives with `agent-(py2py3|py3)` in their name. However, with the `--run` option you can decide to only run the directives which match the regex, but there's no way to exclude only the directives which match. Since negative lookaheads are also not available in Go regexes, there's no way here to exclude something easily.
The other way to do this would be to explicitly list all other directives, so that only them are run, but that could lead to surprises in the future if someone adds a directive.

KSerrania committed:
Add parenthesis in condition

KSerrania committed:
This is ruby, not python

hush-hush commented on pkg/config/render_config.go:
```diff
@@ -64,6 +65,28 @@ func mkContext(buildType string) context {
 			Kubelet:           true,
 			KubeApiServer:     true, // TODO: remove when phasing out from node-agent
 		}
+	case "agent-py2py3":
```
Could we avoid duplicating the whole context for 1 variable ?

hush-hush commented on tasks/agent.py:
```diff
@@ -79,7 +79,7 @@
           puppy=False, development=True, precompile_only=False, skip_assets=False,
           embedded_path=None, rtloader_root=None, python_home_2=None, python_home_3=None,
-          arch='x64'):
+          with_both_python=False, arch='x64'):
```
We could use `python_home_2` and `python_home_3` instead of adding a new variable.

KSerrania replied:
I don't think so: according to the `datadog-agent` software definition, on Windows we never provide these arguments (see [here](https://github.com/DataDog/datadog-agent/pull/4013/files/b160c624b5a92ce06bc02b90f5dcae19d03322b2#diff-26a7fe78c368e14af12bd897cae6b525R63)), and on Linux/MacOS we always provide both arguments (see [here](https://github.com/DataDog/datadog-agent/pull/4013/files/b160c624b5a92ce06bc02b90f5dcae19d03322b2#diff-26a7fe78c368e14af12bd897cae6b525R68)), regardless of whether we built `rtloader` for Python 2, Python 3 or both.

hush-hush replied:
ho yeah ... That's because on Windows the Pythons location are dynamic

hush-hush commented on cmd/agent/main.go:
```diff
@@ -5,7 +5,8 @@
 // +build !windows,!android
 
-//go:generate go run ../../pkg/config/render_config.go agent ../../pkg/config/config_template.yaml ./dist/datadog.yaml
+//go:generate go run ../../pkg/config/render_config.go agent-py2py3 ../../pkg/config/config_template.yaml ./dist/datadog.yaml
```
I'm missing the point here: why do we generate the configuration twice ? Won't the second call overwrite the file ?

KSerrania replied:
See my [previous comment](https://github.com/DataDog/datadog-agent/pull/4013/files/b160c624b5a92ce06bc02b90f5dcae19d03322b2#r313784940): the goal is to selectively generate one command or the other depending on which python versions we are including.
The solution I have right now is far from perfect (it generates the file three times), but keeps the current behaviour (all `go:generate` directives put in the `cmd/agent` folder are automatically run without needing to change something). Another solution would be to explicitly list all directives that would be run, but that means each time someone wants to add / remove a directive, they must add it to the list of generated directives.

Either solution is fine to me, but WDYT?

olivielpeau replied:
how many other `go:generate` directives are run when invoke runs `go generate -x <REPO_PATH>/cmd/agent`?

Depending on how many are run, we could either:

* if there aren't many, stop using `go generate` and simply make the invoke task execute the `go run render_config.go` as many times as needed, with the right arguments
* if there are many, only remove the generation of `datadog.yaml` from `go:generate`, and run this generation only with invoke directly (`go run render_config.go [...]`

Overall I don't feel the feature for which we use `go:generate` justify the extra complexity you're adding here.

KSerrania committed:
Do not define context twice

olivielpeau reviewed (commented):
Left a comment on your technical solution. Please sync with Containers about how what would be acceptable to document (or not) in the docker image, as that may change the technical solution you implement here.

Also, I wasn't clear in my original comment, but I'm OK with just documenting the config option for 6.14 without additional logic, and plan for later how to handle not documenting the config option when only one python version is shipped.

KSerrania committed:
Simpler solution

KSerrania committed:
Remove cross-compiling bits when rendering config

xlucas commented on tasks/agent.py:
```diff
@@ -152,16 +152,24 @@ def build(ctx, rebuild=False, race=False, build_include=None, build_exclude=None
-    ctx.run(cmd.format(REPO_PATH), env=env)
+
+    # Render the Agent configuration file template
+    if with_both_python:
```
A nitpick here but could we use the same as before so that we don't duplicate args entirely ?

```python
cmd = "go run {go_file} {build_type} {template_path} {output_file}"

args = {
    "build_type": "agent-py2py3" if with_both_python else "agent-py3",
    "go_file": "./pkg/config/render_config.go",
    "output_file": "./cmd/agent/dist/datadog.yaml",
    "template_file": "./pkg/config/config_template.yaml",
}

ctx.run(cmd.format(**args), env=env)
```

olivielpeau reviewed (dismissed):
I agree with @xlucas' comment, other than that LGTM

KSerrania committed:
Address review feedback

albertvaka commented on tasks/agent.py:
```diff
@@ -152,16 +152,28 @@ def build(ctx, rebuild=False, race=False, build_include=None, build_exclude=None
+
+    args = {
+        "go_file": "./pkg/config/render_config.go",
+        "build_type": "agent-py2py3" if with_both_python else "agent-py3",
```
Is `agent-py3` also used for when there is only python2? If it's the case, maybe find a better name?

albertvaka reviewed (commented):
Small nit.

albertvaka reviewed (approved):
LGTM. We are not going to build a python2-only image so my previous comment shouldn't be a concern.

xlucas reviewed (approved):
LGTM
